numpy/_core/src/umath/string_buffer.h

```diff
@@ -1428,5 +1428,78 @@ string_replace(Buffer<enc> buf1, Buffer<enc> buf2, Buffer<enc> buf3, npy_int64 c
 }
 
 
+template <ENCODING enc>
+static inline npy_intp
+string_expandtabs_length(Buffer<enc> buf, npy_int64 tabsize)
+{
+    size_t len = buf.num_codepoints();
+
+    npy_intp new_len = 0, line_pos = 0;
+
+    Buffer<enc> tmp = buf;
+    for (size_t i = 0; i < len; i++) {
+        npy_ucs4 ch = *tmp;
+        if (ch == '\t') {
+            if (tabsize > 0) {
+                npy_intp incr = tabsize - (line_pos % tabsize);
+                if (new_len > PY_SSIZE_T_MAX - incr) {
+                    npy_gil_error(PyExc_OverflowError, "new string is too long");
+                    return -1;
+                }
+                line_pos += incr;
+                new_len += incr;
+            }
+        }
+        else {
+            if (new_len > PY_SSIZE_T_MAX - 1) {
+                npy_gil_error(PyExc_OverflowError, "new string is too long");
+                return -1;
+            }
+            line_pos++;
+            new_len++;
+            if (ch == '\n' || ch == '\r') {
+                line_pos = 0;
+            }
+        }
+        tmp++;
+    }
+    return new_len;
+}
+
+
+template <ENCODING enc>
+static inline void
+string_expandtabs(Buffer<enc> buf, npy_int64 tabsize, Buffer<enc> out)
+{
+    size_t len = buf.num_codepoints();
+
+    npy_intp new_len = 0, line_pos = 0;
+
+    Buffer<enc> tmp = buf;
+    for (size_t i = 0; i < len; i++) {
+        npy_ucs4 ch = *tmp;
+        if (ch == '\t') {
+            if (tabsize > 0) {
+                npy_intp incr = tabsize - (line_pos % tabsize);
+                line_pos += incr;
+                new_len += incr;
+                out.buffer_memset((npy_ucs4) ' ', incr);
+                out += incr;
+            }
+        }
+        else {
+            line_pos++;
+            new_len++;
+            out.buffer_memset(ch, 1);
+            out++;
+            if (ch == '\n' || ch == '\r') {
+                line_pos = 0;
+            }
+        }
```

Comment:
I guess it depends on what the input looks like, but I suspect for common real-world uses that instead of writing each non-tab character out, it would be faster to keep track of the last non-tab index and instead `buffer_memcpy` between there and here whenever we encounter a tab or the end of the string.

Reply:
With that, the code would become rather like `replace`, except with a more variable replacement string. Should be possible to basically copy the loop from there to implement the faster copying.

Reply:
Agreed, it'd most probably be faster. Let's merge this to get into NumPy 2.0 and I'll open a follow-up issue.